src/main/kotlin/org/rust/ide/hints/RsChainMethodTypeHintsProvider.kt

```diff
@@ -0,0 +1,115 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.hints
+
+import com.intellij.codeInsight.hints.*
+import com.intellij.codeInsight.hints.presentation.InlayPresentation
+import com.intellij.codeInsight.hints.presentation.InsetPresentation
+import com.intellij.codeInsight.hints.presentation.MenuOnClickPresentation
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.DumbService
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.ui.layout.panel
+import org.rust.ide.utils.isEnabledByCfg
+import org.rust.lang.RsLanguage
+import org.rust.lang.core.psi.RsDotExpr
+import org.rust.lang.core.psi.RsMethodCall
+import org.rust.lang.core.psi.ext.RsElement
+import org.rust.lang.core.psi.ext.childOfType
+import org.rust.lang.core.psi.ext.endOffset
+import org.rust.lang.core.psi.ext.parentDotExpr
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.type
+import javax.swing.JComponent
+
+class RsChainMethodTypeHintsProvider : InlayHintsProvider<RsChainMethodTypeHintsProvider.Settings> {
+    override val key: SettingsKey<Settings> get() = KEY
+
+    override val name: String get() = "Chain method hints"
+
+    override val previewText: String
```

Comment:
Preview doesn't work because our type inference doesn't work properly for dummy files.
I think we can omit preview for now, make some changes in type inference and after that enable it